controllers/cache/scale_down.go

```diff
@@ -0,0 +1,271 @@
+package cache
+
+import (
+	"errors"
+	"fmt"
+	rediscli "github.com/go-redis/redis"
+	"github.com/goharbor/harbor-cluster-operator/lcm"
+	redisCli "github.com/spotahome/redis-operator/api/redisfailover/v1"
+	corev1 "k8s.io/api/core/v1"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"reflect"
+)
+
+// DownScale reconcile will downscale Redis sentinel cluster if replicas downscale.
+// It does:
+// - check resource have been updated, if updated priority to perform rolling upgrade
+// - get a list of nodes that need to be leave
+// - get redis server and sentinel pod list
+// - get redis server password
+// - check whether the leave node is the master node, if is a master node, needs to manual failover
+func (redis *RedisReconciler) ScaleDownCache() (*lcm.CRStatus, error) {
+
+	crdClient := redis.DClient.WithResource(redisFailoversGVR).WithNamespace(redis.Namespace)
+	if redis.ExpectCR == nil {
+		return cacheUnknownStatus(), nil
+	}
+
+	var actualCR redisCli.RedisFailover
+	var expectCR redisCli.RedisFailover
+
+	if err := runtime.DefaultUnstructuredConverter.
+		FromUnstructured(redis.ActualCR.UnstructuredContent(), &actualCR); err != nil {
+		return cacheNotReadyStatus(DefaultUnstructuredConverterError, err.Error()), err
+	}
+
+	if err := runtime.DefaultUnstructuredConverter.
+		FromUnstructured(redis.ExpectCR.UnstructuredContent(), &expectCR); err != nil {
+		return cacheNotReadyStatus(DefaultUnstructuredConverterError, err.Error()), err
+	}
+
+	if isUpgradeResource(&expectCR, &actualCR) {
+		redis.Log.Info("Resources have been updated, priority to perform rolling upgrade",
+			"namespace", redis.Namespace, "name", redis.Name)
+	}
+
+	expectReplica := expectCR.Spec.Redis.Replicas
+	actualReplica := actualCR.Spec.Redis.Replicas
+
+	downscales := calculateDownscales(expectCR, actualCR)
+	leavingNodes := leavingNodeNames(downscales)
+
+	if len(leavingNodes) != 0 {
+		redis.Log.Info("Migrating data away from nodes", "nodes", leavingNodes)
+	}
+
+	if expectReplica < actualReplica {
+
+		_, redisPodList, err := redis.GetStatefulSetPods()
```

Comment:
My question why we need to do such lower operations to do scaling of redis? Does the Redis CR not supporting updating process? e.g.: we just change the replicas of the Redis server/sentinel servers and the redis CR reconcile will help to do the work.